module3/homework/passwordCheck/validation.cpp

```diff
@@ -1,2 +1,55 @@
 #include "validation.hpp"
-// TODO: Put implementations here
\ No newline at end of file
+#include <algorithm>
+#include <string>
+
+std::string getErrorMessage(ErrorCode errorNumber) {
+    switch (errorNumber) {
+        case ErrorCode::Ok: {
+            return "OK";
+        }
+        case ErrorCode::PasswordNeedsAtLeastNineCharacters: {
+            return "Password needs at least nine characters";
+        }
+        case ErrorCode::PasswordNeedsAtLeastOneNumber: {
+            return "Password needs at least one number";
+        }
+        case ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter: {
+            return "Password needs at least one special character";
+        }
+        case ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter: {
+            return "Password needs at least one uppercase letter";
+        }
+        case ErrorCode::PasswordsDoesNotMatch: {
+            return "Passwords does not match";
+        }
+        default: {
+            return "Unknown error";
+        }
+    }
+}
+
+bool doesPasswordsMatch(const std::string& password, const std::string& repeatedPassword) {
+    return (password == repeatedPassword);
+}
+
+ErrorCode checkPasswordRules(const std::string& password) {
+    if (password.size() < 9) {
+        return ErrorCode::PasswordNeedsAtLeastNineCharacters;
+    } else if (!std::any_of(password.begin(), password.end(), ::isdigit)) {
+        return ErrorCode::PasswordNeedsAtLeastOneNumber;
+    } else if (!std::any_of(password.begin(), password.end(), ::ispunct)) {
```

Comment:
I am not sure what ::ispunct will find, will it recognize sth like: <>/_ (which are not punct characters)?
I found a bit different solution:
```suggestion
    } else if (std::all_of(password.begin(), password.end(), ::isalnum)) {
```

You can decide what to choose

Reply:
ispunct is for special characters: !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~ while isalnum is for checking digits, upper and lowercase letters

Reply:
Ok, so that will be good - lets stay with ::ispunct. 
It looks that I found different description of ::ispunct. 
In my proposition it was checking if all of the elements are alpha numeric what is also incorrect and means that there is no special character.